### PR TITLE
Add declarative config health datastore

### PR DESCRIPTION
**Pull request by ivan-degtiarenko**
## Description

In the early stages of developing the declarative configuration feature, we decided to reuse `integrationhealth` datastore and API to track the health of the declarative config. This proved to have some cons:
* No ability to add resource name to API as there's no such field in integration health proto
* Possibility of clashing IDs  - for example, if a resource is notifier: https://github.com/stackrox/stackrox/pull/6172#discussion_r1206101388

This PR introduces a separate datastore for declarative config health datastore. 

What this PR does not include(on purpose):
* Usage of new datastore in service or integration health reporter
* UI/tests adjustments
* Migration of existing declarative config health to a new datastore

## Checklist
- [ ] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] Evaluated and added CHANGELOG entry if required
- [ ] Determined and documented upgrade steps
- [ ] Documented user-facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))

If any of these don't apply, please comment below.

## Testing Performed

1. Unit test is added

### CODE REVIEW

File: proto/storage/declarative_config_health.proto
Comment (dhaus67) [inline proto/storage/declarative_config_health.proto:21]: In addition to the resource name, we can also think about adding a type here to infer a) which resource this is to potentially create a link to the resource within the UI b) distinguish between config map and other objects.

WDYT?

File: proto/storage/declarative_config_health.proto
Comment (dhaus67) [inline proto/storage/declarative_config_health.proto:17]: Let's also add type UUID here now.

File: central/declarativeconfig/health/datastore.go
Comment (dhaus67) [inline central/declarativeconfig/health/datastore.go:17]: Let's see when we refactor the existing code to use the new datastore, but this could be potentially unused since in the current implementation it's also not used.

File: central/declarativeconfig/health/datastore_impl.go
Comment (dhaus67) [inline central/declarativeconfig/health/datastore_impl.go:24]: (completely unrelated with the PR):
we should probably think about creating another util function in SAC code similar to `VerifyAuthZOK` which doesn't return an error in case `ok == false`, and document it properly so that each API behaves the same (on GET return no `401`, on other methods return `401` in case no permissions).

File: central/declarativeconfig/health/datastore_impl.go
Comment (dhaus67) [inline central/declarativeconfig/health/datastore_impl.go:55]: Let's be consistent on the usage of either `%s` or `%q`, I don't have a strong preference to either (albeit it seems that `%q` is what we mostly use for IDs in APIs), but it should be consistent for this datastore for sure.

File: central/declarativeconfig/health/datastore_impl_test.go
Comment (dhaus67) [inline central/declarativeconfig/health/datastore_impl_test.go:80]: Somehow the comments don't match, in the update case we receive an error 🤔

File: central/declarativeconfig/health/datastore_impl_test.go
Comment (dhaus67) [inline central/declarativeconfig/health/datastore_impl_test.go:87]: Same as above.

File: central/declarativeconfig/health/datastore_impl_test.go
Comment (dhaus67) [inline central/declarativeconfig/health/datastore_impl_test.go:134]:
```suggestion
	s.ErrorIs(err, errox.NotFound)
```

**Review by dhaus67 — COMMENTED** [review]
When first introducing a new datastore (and subsequently a new table) you'd also have to add a migration (see [this as example](https://github.com/stackrox/stackrox/blob/master/migrator/migrations/m_175_to_m_176_create_notification_schedule_table/migration.go)).

We should add this within this PR as well.

File: central/declarativeconfig/health/datastore_impl_test.go
Comment (ivan-degtiarenko) [reply central/declarativeconfig/health/datastore_impl_test.go:1]: I ran them manually - didn't fix the Makefile since https://github.com/stackrox/stackrox/pull/6245 once merged should fix the issue

File: central/declarativeconfig/health/datastore.go
Comment (ivan-degtiarenko) [reply central/declarativeconfig/health/datastore.go:17]: True - let's remove it after we implement logic on top of this PR

File: migrator/migrations/m_183_to_m_184_create_declarative_config_health_table/migration_test.go
Comment (dhaus67) [inline migrator/migrations/m_183_to_m_184_create_declarative_config_health_table/migration_test.go:9]: Wrong UUID package, we should use the stackrox one (this also leads to the go.mod change which is unwanted).

**Review by dhaus67 — APPROVED** [review]
One comment regarding the wrong UUID package being used, otherwise LGTM.

Probably worthwhile to wait for the approval from data sheperds (probably from David 😄 ) to check everything related to the migration.

File: central/declarativeconfig/health/datastore_impl.go
Comment (dashrews78) [inline central/declarativeconfig/health/datastore_impl.go:16]: General question on the SAC checks.  SAC is also checked in the store.  Are the checks the same or are the checks here at a different level than the ones in the store?  Basically I'm just asking if we are checking the same SAC twice.

File: migrator/migrations/m_183_to_m_184_create_declarative_config_health_table/migration.go
Comment (dashrews78) [inline migrator/migrations/m_183_to_m_184_create_declarative_config_health_table/migration.go:1]: Is the migration only creating the table and not populating it from some other source?  We only need a migration if we are going to populate the data.  The table resources should be created without the need for a migration.  Obviously you should run an upgrade to confirm that.

Follow up question, where is the data now and does the data need to be moved into this new table?

**Review by dashrews78 — CHANGES_REQUESTED** [review]
Couple of questions on SAC and the migration.

File: proto/storage/declarative_config_health.proto
Comment (dashrews78) [reply proto/storage/declarative_config_health.proto:31]: Do we envision any use cases where we may query on some of these other fields to avoid `walks` or is this data sufficiently small so the `walks` are fine.

File: central/declarativeconfig/health/datastore_impl.go
Comment (ivan-degtiarenko) [reply central/declarativeconfig/health/datastore_impl.go]: That's a very good observation - I removed duplicating checks from `datastore_impl`

File: migrator/migrations/m_183_to_m_184_create_declarative_config_health_table/migration.go
Comment (ivan-degtiarenko) [reply migrator/migrations/m_183_to_m_184_create_declarative_config_health_table/migration.go:1]: This migration only creates the table - we took inspiration from this PR which adds similar migration for notificationschedules table: https://github.com/stackrox/stackrox/pull/5067

However, we would need to migrate data from `integrationhealth` store to this one(more specifically, all objects with `DECLARATIVE_CONFIG` type).

Currently, I'm a little bit unsure how it is supposed to work - I see a few options:
1. We create two separate migrations - `create table` migration and `move data to table` migration
2. We create one migration - `create table + move data to table` migration
3. `create table` migration is not needed, only `move data to table` migration

Comment (dashrews78) [reply migrator/migrations/m_183_to_m_184_create_declarative_config_health_table/migration.go:1]: Ah, well my comment there applies.  But since then my thinking has evolved.  

Option 3 is not valid because what postgres things that aren't created by a migration are not created until after the migrations.  My preference is option 2 for a couple of reasons.  It reduces our migrations and in reality it doesn't matter whether the table is created or not until you try to use it.

For my own education, declarative config was added in 4.0?

Comment (dhaus67) [reply migrator/migrations/m_183_to_m_184_create_declarative_config_health_table/migration.go:1]: >For my own education, declarative config was added in 4.0?
It was added as a developer preview, so I guess we don't have to be as strict here w.r.t notifying people in the CHANGELOG or something like this. 4.1 will be the GA release of declarative config.
 
Just to know this myself: the existing migrations we have where we added empty tables essentially should have never been migrations in the first place, right?

Comment (ivan-degtiarenko) [reply migrator/migrations/m_183_to_m_184_create_declarative_config_health_table/migration.go]: > Would there potentially be any declarative config data existing in customer environments?

Yes. However, AFAIU this functionality was presented only to a few customers who expressed interest in experimenting with this feature

Comment (dashrews78) [reply migrator/migrations/m_183_to_m_184_create_declarative_config_health_table/migration.go]: got it.  To be safe then when you hook up the services to the store you should probably migrate the data

**Review by dashrews78 — APPROVED** [review]
LGTM.  Thanks for addressing my comments/questions.

File: proto/storage/declarative_config_health.proto
Comment (dhaus67) [reply proto/storage/declarative_config_health.proto:31]: I suppose this data should be sufficiently small (currently within the MVP we don't not more than 100 configs), it'd be more of a future-type-thing if that makes sense.

Comment (dashrews78) [reply proto/storage/declarative_config_health.proto:31]: Sounds fair.  that is something I'm looking at for what I'm working on so I figured I would ask since it is fresh in my mind.  Easy enough to promote something to a column for search when the need arises.

**Pull request merged**
